From 0e3ce049b25746e478a854f5d876ee26c11f8339 Mon Sep 17 00:00:00 2001
Message-Id: <0e3ce049b25746e478a854f5d876ee26c11f8339.1413837856.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: Daisy Sun <daisy.sun@intel.com>
Date: Thu, 16 Jan 2014 20:03:33 -0800
Subject: [PATCH 062/312] MUST_REBASE [VPG]: drm/i915: DPST signal sent by
 kill_sig_info

When DPST interrupt is generated, kernel will send this info to
user-mode via signal.
Change the function which send the signal from send_sig_info() to
kill_pid_info(), the latter has more sanity check, it's safer
if the receiver could get killed.

Issue: APDEV-1074
Change-Id: I3aeb412883790a9c2a428b17524ab0f1c96c4e9f
Signed-off-by: Daisy Sun <daisy.sun@intel.com>
Reviewed-on: https://android.intel.com/162722
Tested-by: sys_abtbuild <sys_abtbuild@intel.com>
Reviewed-by: Azad, Vinit <vinit.azad@intel.com>
---
 drivers/gpu/drm/i915/i915_dpst.c |   16 ++++++++++++----
 drivers/gpu/drm/i915/i915_drv.h  |    2 +-
 2 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_dpst.c b/drivers/gpu/drm/i915/i915_dpst.c
index 5828a56..6265215 100644
--- a/drivers/gpu/drm/i915/i915_dpst.c
+++ b/drivers/gpu/drm/i915/i915_dpst.c
@@ -333,6 +333,7 @@ i915_dpst_init(struct drm_device *dev,
 	struct drm_crtc *crtc;
 	struct drm_display_mode *mode = NULL;
 	u32 blm_hist_guard, gb_val;
+	struct pid *cur_pid;
 
 	/* Get information about current display mode */
 	crtc = intel_get_crtc_for_pipe(dev, PIPE_A);
@@ -349,9 +350,12 @@ i915_dpst_init(struct drm_device *dev,
 	}
 
 	/* Store info needed to talk to user mode */
-	dev_priv->dpst.task = current;
+	cur_pid = get_task_pid(current, PIDTYPE_PID);
+	put_pid(dev_priv->dpst.pid);
+	dev_priv->dpst.pid = cur_pid;
 	dev_priv->dpst.signal = ioctl_data->init_data.sig_num;
 
+
 	/* Setup guardband delays and threshold */
 	blm_hist_guard = I915_READ(BLM_HIST_GUARD);
 	blm_hist_guard |= (ioctl_data->init_data.gb_delay << 22)
@@ -409,9 +413,13 @@ i915_dpst_irq_handler(struct drm_device *dev)
 	struct drm_i915_private *dev_priv = dev->dev_private;
 
 	/* Notify user mode of the interrupt */
-	if (dev_priv->dpst.task != NULL)
-		send_sig_info(dev_priv->dpst.signal, SEND_SIG_FORCED,
-							dev_priv->dpst.task);
+	if (dev_priv->dpst.pid != NULL) {
+		if (kill_pid_info(dev_priv->dpst.signal, SEND_SIG_FORCED,
+							dev_priv->dpst.pid)) {
+			put_pid(dev_priv->dpst.pid);
+			dev_priv->dpst.pid = NULL;
+		}
+	}
 }
 
 int
diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 0eea1bb..01fc8b1 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -1467,7 +1467,7 @@ struct drm_i915_private {
 
 	/* DPST information */
 	struct {
-		struct task_struct *task;
+		struct pid *pid;
 		u32 signal;
 		u32 blc_adjustment;
 		bool user_enable;	/* user client wishes to enable */
-- 
1.7.9.5

